File: backend/app/schemas/workflow.py
from typing import List, Dict, Any
from pydantic import BaseModel, field_validator
from ..nodes.factory import NodeFactory


class WorkflowNode(BaseModel):
    """
    A node represents a single step in a workflow.
    """

    id: str  # ID in the workflow
    node_type: str  # Name of the node type
    config: Dict[str, Any] = {}  # Configuration parameters

    @field_validator("node_type")
    def type_must_be_in_factory(cls, v: str):
        if not NodeFactory.is_valid_node_type(v):
            raise ValueError(f"Node type '{v}' is not valid.")
        return v


class WorkflowLink(BaseModel):
    """
    A link connects an output key of a source node to an input key of a target node.
    """

    source_id: str
    source_output_key: str
    target_id: str
    target_input_key: str


class Workflow(BaseModel):
    """
    A workflow is a DAG of nodes.
    """

    nodes: List[WorkflowNode]
    links: List[WorkflowLink]

File: frontend/src/types/workflow.ts
export interface WorkflowNodeCoordinates {
  x: number;
  y: number;
}

export interface WorkflowNode {
  id: string;
  title?: string;
  node_type: string;
  config: Record<string, any>;
  coordinates?: WorkflowNodeCoordinates;
}

export interface WorkflowLink {
  source_id: string;
  source_output_key: string;
  target_id: string;
  target_input_key: string;
}

export interface WorkflowDefinition {
  nodes: WorkflowNode[];
  links: WorkflowLink[];
  test_inputs: Record<string, any>[];
}

export interface Workflow {
  id: string;
  key?: string;
  name: string;
  description: string;
  definition: WorkflowDefinition;
  created_at: string; // ISO datetime string
  updated_at: string; // ISO datetime string
}

export interface Template {
  file_name: string;
  name: string;
  description: string;
  features: string[];
}

export interface Dataset {
  id: string;
  name: string;
  description: string;
}